File: src/sci/impl/java/options.clj
(ns sci.impl.java.options
  (:require [sci.impl.java.options.namespace]))

(set! *warn-on-reflection* true)

;; Two step process because we need the class in the return types of `addBinding`
;; and `addToNamespace`. See https://stackoverflow.com/questions/29329798/clojure-gen-class-returning-own-class.

(gen-class
 :name borkdude.sci.options.Options
 :constructors {[] []}
 :state "state"
 :init "init"
 :prefix "opts-")

(gen-class
 :name borkdude.sci.options.Options
 :constructors {[] []}
 :state "state"
 :init "init"
 :prefix "opts-"
 :methods [["addBinding" [java.lang.String java.lang.Object] borkdude.sci.options.Options]
           ["addNamespace" [borkdude.sci.options.Namespace] borkdude.sci.options.Options]
           ["deny" [java.lang.String] borkdude.sci.options.Options]
           ["allow" [java.lang.String] borkdude.sci.options.Options]
           ["val" [] java.lang.Object]])

(defn opts-init []
  [[] (atom {})])

(defn opts-addBinding [^borkdude.sci.options.Options this k v]
  (let [state (.state this)]
    (swap! state assoc-in [:bindings (symbol k)] v))
  this)

(defn opts-deny [^borkdude.sci.options.Options this name]
  (let [state (.state this)]
    (swap! state update :deny (fnil conj #{}) (symbol name)))
  this)

(defn opts-allow [^borkdude.sci.options.Options this name]
  (let [state (.state this)]
    (swap! state update :allow (fnil conj #{}) (symbol name)))
  this)

(defn opts-addNamespace [^borkdude.sci.options.Options this ^borkdude.sci.options.Namespace ns]
  (let [state (.state this)
        ns-val (.val ns)
        ns-name (:name ns-val)
        vars (:vars ns-val)]
    (swap! state assoc-in [:namespaces ns-name] vars))
  this)

(defn opts-val [^borkdude.sci.options.Options this]
  @(.state this))

File: src-java/borkdude/sci/options/Options.java
package borkdude.sci.options;

import java.util.HashMap;
import java.util.ArrayList;
import borkdude.sci.options.Namespace;

public class Options {

    private HashMap<String,Object> _bindings = new HashMap<String,Object>();
    private HashMap<String,Namespace> _namespaces = new HashMap<String, Namespace>();
    private ArrayList<String> _allow = new ArrayList<String>();
    private ArrayList<String> _deny = new ArrayList<String>();
    private String _preset;

    public Options() {
    }

    public Options addBinding(String name, Object value) {
        _bindings.put(name, value);
        return this;
    }

    public Options deny(String name) {
        _deny.add(name);
        return this;
    }

    public Options allow(String name) {
        _allow.add(name);
        return this;
    }

    public Options addNamespace(Namespace n) {
        _namespaces.put(n.getName(), n);
        return this;
    }

    public Options setPreset(String name) {
        _preset = name;
        return this;
    }

    /**
     * Implementation detail, don't use.
     * @return: irrelevant.
     */
    public HashMap<String, Object> _val() {
        HashMap<String, Object> ret = new HashMap<String, Object>();
        ret.put("namespaces", _namespaces);
        ret.put("bindings", _bindings);
        ret.put("allow", _allow);
        ret.put("deny", _deny);
        ret.put("preset", _preset);
        return ret;
    }
}
